documentation/specs/src/interoperability/ethereum-bridge/ethereum_smart_contracts.md

```diff
@@ -1,27 +1,180 @@
+spendable by the current _Bridge_ contract. When ERC20 tokens are transferred
+from Ethereum to Namada, they must be deposited to the _Vault_ via the _Bridge_
+contract.
+
+The _Vault_ contract is fixed forever once the bridge has been deployed.
+
+## Namada-side configuration
+
+When an account on Namada becomes a validator, they must provide two Ethereum
+secp256k1 keys:
+
+- a "hot" key for normal operations
+- a "cold" key for exceptional operations, like emergency withdrawal from the
+  bridge
+
+These keys are used to control the bridge smart contracts, via signing of
+messages. Validators may be challenged periodically to prove they still retain
+knowledge of these keys.
+
+## Deployment
+
+The contracts should be deployable by anyone to any EVM chain using an automated
+script. The following configuration should be agreed up front by Namada
+governance before deployment:
+
+- details of the initial active validator set that will control the bridge -
+  specifically, for each validator:
+  - their hot Ethereum address
+  - their cold Ethereum address
+  - their voting power on Namada for the epoch when the bridge will launch
+- the total supply of the wNAM ERC20 token, which will represent Namada-native
+  NAM on the EVM chain
+- an initial whitelist of ERC20 tokens that may cross the bridge from Ethereum
+  to Namada - specifically, for each whitelisted ERC20:
+    - the Ethereum address of the ERC20 contract
+    - a cap on the total amount that may cross the bridge, in units of ERC20
+
+After a deployment has finished successfully, the deployer must not have any
+privileged control of any of the contracts deployed. Any privileged actions must
+only be possible via a message signed by a validator set that the smart
+contracts are storing details of.
+
+## Communication
+
+### From Ethereum to Namada
+
+A Namada chain's validators are configured to listen to events emitted by the
+smart contracts pointed to by the _Proxy_ contract. The address of the _Proxy_
+contract is set in a governance parameter in Namada storage. Namada validators
+treat emitted events as authoritative and take action on them. Namada also knows
+the address of the _wNAM_ ERC20 contract via a governance parameter, and treats
+transfers of this ERC20 to Namada as an indication to release native NAM from
+the `#EthBridgeEscrow` account on Namada, rather than to mint a wrapped ERC20 as
+is the case with all other ERC20s.
+
+### From Namada to Ethereum
+
+At any time, the _Governance_ and _Bridge_ contracts must store:
+
+- a hash of the current Namada epoch's active validator set
+- a hash of another epoch's active validator set. When the bridge is first
+  deployed, this will also be the current Namada epoch's active validator set,
+  but after the first valset update is submitted to the _Governance_ smart
+  contract, this hash will always be an adjacent Namada epoch's active validator
+  set i.e. either the previous epoch's, or the next epoch's
```

Comment:
This is probably not the clearest way to explain it, and there is a reference here to `valset update` which isn't defined until later in the spec. The idea is at any time we have two hashes, and (at least) one of them is the current validator set's hash. The other hash should be the next validator set's hash most of the time, but after an epoch transition, we'll be in the situation where we have the current + previous validator sets' hashes, usually only briefly until a valset update has been submitted.